drgn_tools/meminfo.py

```diff
@@ -0,0 +1,767 @@
+) -> int:
+    """
+    Get the global stats item from all NUMA nodes at the target key.
+
+    :param prog: drgn program
+    :param key: The target node statistics item's name
+    :param offset: The target node statistics item's offset
+    :returns: The target node statistics item's value
+    """
+    if not is_valid_node_page_state(prog, key):
+        print("INVALID enum item %s in node statistics" % (key))
+        return 0
+
+    try:
+        cache = prog.cache.setdefault("drgn_tools.meminfo", {})
+        vm_node_stat = cache["vm_node_stat"]
+    except KeyError:
+        vm_node_stat = prog["vm_node_stat"].read_()
+        cache["vm_node_stat"] = vm_node_stat
+
+    try:
+        item = int(prog[key]) + offset
+        ret = vm_node_stat[item]
+    except LookupError:
+        raise IndexError("UNKNOWN enum item %d in node statistics" % (item))
+    return max(0, ret.counter.value_())
+
+
+def get_global_zone_page_state(
+    prog: Program, key: str, offset: int = 0
+) -> int:
+    """
+    Get the global stats item from all memory zones at the target key.
+
+    :param prog: drgn program
+    :param key: The target zone statistics item's name
+    :param offset: The target zone statistics item's offset
+    :returns: The target zone statistics item's name
+    """
+    if not is_valid_zone_page_state(prog, key):
+        print("INVALID enum item %s in zone statistics" % (key))
+        return 0
+
+    try:
+        cache = prog.cache.setdefault("drgn_tools.meminfo", {})
+        vm_zone_stat = cache["vm_zone_stat"]
+    except KeyError:
+        if "vm_stat" in prog:
+            vm_zone_stat = prog["vm_stat"].read_()
+        else:
+            vm_zone_stat = prog["vm_zone_stat"].read_()
+        cache["vm_zone_stat"] = vm_zone_stat
+
+    try:
+        item = int(prog[key]) + offset
+        ret = vm_zone_stat[item]
+    except LookupError:
+        raise IndexError("UNKNOWN enum item %d in zone statistics" % (item))
+    return max(0, ret.counter.value_())
+
+
+def get_global_page_state(prog: Program, key: str, offset: int = 0) -> int:
+    """
+    Get either global node page statistics or global zone page statistics.
+    |get_global_zone_page_state| and |get_global_node_page_state| should NOT
+    be directly called from other functions.
+    """
+    if key == "NR_SLAB_RECLAIMABLE" and key not in prog:
+        key = "NR_SLAB_RECLAIMABLE_B"
+    if key == "NR_SLAB_UNRECLAIMABLE" and key not in prog:
+        key = "NR_SLAB_UNRECLAIMABLE_B"
+    if key == "NR_ANON_MAPPED" and key not in prog:
+        key = "NR_ANON_PAGES"
+    if key == "NR_KERNEL_STACK_KB" and key not in prog:
+        key = "NR_KERNEL_STACK"
+
+    if is_valid_zone_page_state(prog, key):
+        return get_global_zone_page_state(prog, key, offset)
+    if is_valid_node_page_state(prog, key):
+        return get_global_node_page_state(prog, key, offset)
+    raise IndexError("INVALID enum item %s in node/zone statistics", key)
```

Comment:
This is definitely amazing. Thanks for your suggestion.

When I implemented this module, I basically followed the way of reading global statistics counters in the kernel: by having functions like: global_node_page_state(xxx) and global_zone_page_state(xxx). Then, I realized that I might be reading ```vm_node_stat``` and ```vm_zone_stat``` many many times, so I decided to cache them.

I agree that we should avoid caching because we'd like to support live vmcores. I like the way done in your suggestion in that we can parse and store all statistics items all at once without further caching them.